sql/plan/indexed_in_subquery_filter.go

```diff
@@ -0,0 +1,156 @@
+// results that should be matched against the `subquery` results; this
+// condition is still checked here because `child` is allowed to
+// return non-matching rows. `equals` true means this node will call
+// `subquery.Eval` and expect a single result, whereas `equals` false
+// means this node will call `subquery.EvalMultiple` and expect 0 or
+// more results.
+func NewIndexedInSubqueryFilter(subquery *Subquery, child sql.Node, padding int, filterField *expression.GetField, equals bool) sql.Node {
+	return &IndexedInSubqueryFilter{subquery, child, padding, filterField, equals}
+}
+
+type IndexedInSubqueryFilter struct {
+	Subquery *Subquery
+	Child    sql.Node
+	Padding  int
+	GetField *expression.GetField
+	Equals   bool
+}
+
+func (i *IndexedInSubqueryFilter) Resolved() bool {
+	return i.Subquery.Resolved() && i.Child.Resolved()
+}
+
+func (i *IndexedInSubqueryFilter) String() string {
+	pr := sql.NewTreePrinter()
+	_ = pr.WriteNode("IndexedInSubqueryFilter(%s IN (%s))", i.GetField, i.Subquery)
+	_ = pr.WriteChildren(i.Child.String())
+	return pr.String()
+}
+
+func (i *IndexedInSubqueryFilter) DebugString() string {
+	pr := sql.NewTreePrinter()
+	_ = pr.WriteNode("IndexedInSubqueryFilter(%s IN (%s))", sql.DebugString(i.GetField), sql.DebugString(i.Subquery))
+	_ = pr.WriteChildren(sql.DebugString(i.Child))
+	return pr.String()
+}
+
+func (i *IndexedInSubqueryFilter) Schema() sql.Schema {
+	return i.Child.Schema()
+}
+
+func (i *IndexedInSubqueryFilter) Children() []sql.Node {
+	return []sql.Node{i.Child}
+}
+
+func (i *IndexedInSubqueryFilter) WithChildren(children ...sql.Node) (sql.Node, error) {
+	if len(children) != 1 {
+		return nil, sql.ErrInvalidChildrenNumber.New(i, len(children), 1)
+	}
+	return NewIndexedInSubqueryFilter(i.Subquery, children[0], i.Padding, i.GetField, i.Equals), nil
+}
+
+func (i *IndexedInSubqueryFilter) RowIter(ctx *sql.Context, row sql.Row) (sql.RowIter, error) {
+	padded := make(sql.Row, len(row)+i.Padding)
+	copy(padded[:], row[:])
+	var res []interface{}
+	var err error
+	if i.Equals {
+		resi, err := i.Subquery.Eval(ctx, padded)
+		if err != nil {
+			return nil, err
+		}
+		res = append(res, resi)
+	} else {
+		res, err = i.Subquery.EvalMultiple(ctx, padded)
+		if err != nil {
+			return nil, err
+		}
+	}
+	tupLits := make([]sql.Expression, len(res))
+	for j := range res {
+		tupLits[j] = expression.NewLiteral(res[j], i.Subquery.Type())
+	}
+	expr := expression.NewInTuple(i.GetField, expression.NewTuple(tupLits...))
+	return NewFilterIter(ctx, expr, &indexedInSubqueryIter{ctx, res, i.Child, nil, 0}), nil
+}
+
+type indexedInSubqueryIter struct {
+	Ctx   *sql.Context
+	Rows  []interface{}
+	Child sql.Node
+	Cur   sql.RowIter
+	I     int
+}
+
+func (i *indexedInSubqueryIter) Next() (sql.Row, error) {
+	var ret sql.Row
+	err := io.EOF
+	for err == io.EOF {
+		if i.Cur == nil {
+			if i.I >= len(i.Rows) {
+				return nil, io.EOF
+			}
+			iter, err := i.Child.RowIter(i.Ctx, sql.NewRow(i.Rows[i.I]))
```

Comment:
should be no need for sql.NewRow here after the above change

Reply:
I don't know...this is a little janky. `EvalMultiple` isn't returning `[]sql.Row`, it's returning `[]interface{}`, where each entry in the slice is a sql engine value. But the call to `Child.RowIter(...)` is expecting a `sql.Row` with a single column which is that value. I think it's typed appropriately as is?

Reply:
Yeah, that's a good point